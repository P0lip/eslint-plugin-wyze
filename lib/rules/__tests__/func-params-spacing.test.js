/* eslint-disable wyze/max-file-length */
import { RuleTester } from 'eslint'
import rule from '../func-params-spacing'
import test from 'ava'

const parserOptions = {
  ecmaVersion: 2017,
  sourceType: 'module',
}

const makeRequiredError = location =>
  `A space is required ${location} '${location === 'after' ? '(' : ')'}'.`
const makeMissingError = location =>
  `There should be no space ${location} '${location === 'after' ? '(' : ')'}'.`

test(() => {
  new RuleTester().run('func-params-spacing', rule, {
    valid: [
      {
        code: 'function foo( a, b, c ) {}',
      },
      {
        code: 'const foo = function( a, b, c ) {}',
        parserOptions,
      },
      {
        code: 'const foo = ( a, b, c ) => {}',
        parserOptions,
      },
      {
        code: 'const foo = ({ a, b, c }) => {}',
        parserOptions,
      },
      {
        code: 'const foo = ([ a, b, c ]) => {}',
        parserOptions,
      },
      {
        code: 'const foo = ({ a }, b, { c }) => {}',
        parserOptions,
      },
      {
        code: 'const foo = ({ a }, b, [ c ]) => {}',
        parserOptions,
      },
      {
        code: 'const foo = ( { a }, b ) => {}',
        parserOptions,
      },
      {
        code: 'const foo = ( b, [ c ] ) => {}',
        parserOptions,
      },
      {
        code: 'const foo = () => {}',
        parserOptions,
      },
      {
        code: 'const foo = a => {}',
        parserOptions,
      },
      {
        code: 'function foo() {}',
        parserOptions,
      },
      {
        code: 'const foo = function() {}',
        parserOptions,
      },
      {
        code:
        `
        const a = ({
          b,
          c,
        }) => {}
        `,
        parserOptions,
      },
      {
        code:
        `
        const keys = properties
          .filter(prop => !prop.computed)
          .map(prop => prop.key.name)
        `,
        parserOptions,
      },
      {
        code:
        `
        const keys = properties
          .filter(function( a ) { return a })
        `,
        parserOptions,
      },
      {
        code: 'const foo = ({ a }: { a: string }): void => {}',
        parser: 'babel-eslint',
        parserOptions,
      },
      {
        code: 'const foo = ( a: string ): void => {}',
        parser: 'babel-eslint',
        parserOptions,
      },
      {
        code: 'const foo = ( { a }: Object ): void => {}',
        parser: 'babel-eslint',
        parserOptions,
      },
    ],
    invalid: [
      {
        code: 'function foo( a, b, c) {}',
        errors: [
          makeRequiredError('before'),
        ],
      },
      {
        code: 'function foo(a, b, c) {}',
        errors: [
          makeRequiredError('after'),
          makeRequiredError('before'),
        ],
      },
      {
        code: 'const foo = function(a, b, c ) {}',
        errors: [
          makeRequiredError('after'),
        ],
        parserOptions,
      },
      {
        code: 'const foo = ( a, b, c) => {}',
        errors: [
          makeRequiredError('before'),
        ],
        parserOptions,
      },
      {
        code: 'const foo = (a, b, c) => {}',
        errors: [
          makeRequiredError('after'),
          makeRequiredError('before'),
        ],
        parserOptions,
      },
      {
        code: 'const foo = ({ a } ) => {}',
        errors: [
          makeMissingError('before'),
        ],
        parserOptions,
      },
      {
        code: 'const foo = ( { a } ) => {}',
        errors: [
          makeMissingError('after'),
          makeMissingError('before'),
        ],
        parserOptions,
      },
      {
        code: 'const foo = ({ a }, b ) => {}',
        errors: [
          makeRequiredError('after'),
        ],
        parserOptions,
      },
      {
        code: 'const foo = ( { a }, b, [ c ] ) => {}',
        errors: [
          makeMissingError('after'),
          makeMissingError('before'),
        ],
        parserOptions,
      },
      {
        code:
        `
        const foo = ( {
          a
        }) => {}
        `,
        errors: [
          makeMissingError('after'),
        ],
        parserOptions,
      },
      {
        code:
        `
        const keys = properties
          .filter( prop => !prop.computed )
          .map( prop => prop.key.name )
        `,
        errors: [
          makeMissingError('after'),
          makeMissingError('before'),
          makeMissingError('after'),
          makeMissingError('before'),
        ],
        parserOptions,
      },
      {
        code: 'const foo = ( ) => {}',
        errors: [
          makeMissingError('after'),
        ],
        parserOptions,
      },
      {
        code: 'const foo = (a: string ) => {}',
        errors: [
          makeRequiredError('after'),
        ],
        parser: 'babel-eslint',
        parserOptions,
      },
    ],
  })
})

