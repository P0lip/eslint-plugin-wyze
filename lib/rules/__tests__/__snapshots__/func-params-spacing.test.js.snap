// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`func-params-spacing applies fixes to callbacks 1`] = `
Object {
  "errorCount": 0,
  "results": Array [
    Object {
      "errorCount": 0,
      "filePath": "/Users/neil/Projects/personal/eslint-plugin-wyze/lib/rules/__tests__/fixtures/func-params-spacing/callbacks.js",
      "messages": Array [],
      "output": "
const a = b.reduce(( c, d ) => c + d, 0)
const a = b.reduce(( c, d ) => c + d, 0)
const a = b.reduce(( c, d ) => c + d, 0)
const a = b.reduce(() => Math.random(5), 0)
const a = b.reduce(function ( c, d ) { c + d }, 0)
",
      "warningCount": 0,
    },
  ],
  "warningCount": 0,
}
`;

exports[`func-params-spacing applies fixes to classes 1`] = `
Object {
  "errorCount": 0,
  "results": Array [
    Object {
      "errorCount": 0,
      "filePath": "/Users/neil/Projects/personal/eslint-plugin-wyze/lib/rules/__tests__/fixtures/func-params-spacing/classes.js",
      "messages": Array [],
      "output": "
class Component {
  componentWillMount() {
    // Do something...
  }
}

class Component {
  componentWillMount( a ) {
    // Do something...
  }
}

class Component {
  componentWillMount( a ) {
    // Do something...
  }
}
",
      "warningCount": 0,
    },
  ],
  "warningCount": 0,
}
`;

exports[`func-params-spacing applies fixes to destructuring params 1`] = `
Object {
  "errorCount": 0,
  "results": Array [
    Object {
      "errorCount": 0,
      "filePath": "/Users/neil/Projects/personal/eslint-plugin-wyze/lib/rules/__tests__/fixtures/func-params-spacing/destructuring.js",
      "messages": Array [],
      "output": "
const foo = ({ a }) => {}
const foo = ({ a }) => {}
const foo = ({ a }) => {}
const foo = ( { a }, b ) => {}
const foo = ( { a }, b ) => {}
const foo = ({ a }, b, [ c ]) => {}
const foo = ({ a }, b, [ c ]) => {}
const foo = ({ a }, b, [ c ]) => {}
",
      "warningCount": 0,
    },
  ],
  "warningCount": 0,
}
`;

exports[`func-params-spacing applies fixes to flow types 1`] = `
Object {
  "errorCount": 0,
  "results": Array [
    Object {
      "errorCount": 0,
      "filePath": "/Users/neil/Projects/personal/eslint-plugin-wyze/lib/rules/__tests__/fixtures/func-params-spacing/flowtype.js",
      "messages": Array [],
      "output": "
const foo = ( a: string ) => {}
const foo = ( a: string ) => {}
const foo = ( a: string ) => {}
const foo = ({ a }: { a: string }): void => {}
const foo = ({ a }: { a: string }): void => {}
const foo = ({ a }: { a: string }): void => {}
const foo = ( { a }: Object ): void => {}
const foo = ( { a }: Object ): void => {}
const foo = ( { a }: Object ): void => {}
",
      "warningCount": 0,
    },
  ],
  "warningCount": 0,
}
`;

exports[`func-params-spacing applies fixes to function declarations 1`] = `
Object {
  "errorCount": 0,
  "results": Array [
    Object {
      "errorCount": 0,
      "filePath": "/Users/neil/Projects/personal/eslint-plugin-wyze/lib/rules/__tests__/fixtures/func-params-spacing/function-declaration.js",
      "messages": Array [],
      "output": "
function foo( a, b, c ) {}
function foo( a, b, c ) {}
function foo( a, b, c ) {}
function foo() {}
",
      "warningCount": 0,
    },
  ],
  "warningCount": 0,
}
`;

exports[`func-params-spacing applies fixes to function expressions 1`] = `
Object {
  "errorCount": 0,
  "results": Array [
    Object {
      "errorCount": 0,
      "filePath": "/Users/neil/Projects/personal/eslint-plugin-wyze/lib/rules/__tests__/fixtures/func-params-spacing/function-expression.js",
      "messages": Array [],
      "output": "
const foo = function( a, b, c ) {}
const foo = function( a, b, c ) {}
const foo = function( a, b, c ) {}
const foo = ( a, b, c ) => {}
const foo = ( a, b, c ) => {}
const foo = ( a, b, c ) => {}
const foo = () => {}
const foo = function() {}
",
      "warningCount": 0,
    },
  ],
  "warningCount": 0,
}
`;

exports[`func-params-spacing applies fixes to multiline functions 1`] = `
Object {
  "errorCount": 0,
  "results": Array [
    Object {
      "errorCount": 0,
      "filePath": "/Users/neil/Projects/personal/eslint-plugin-wyze/lib/rules/__tests__/fixtures/func-params-spacing/multi-line.js",
      "messages": Array [],
      "output": "
const foo = ({
  a
}) => {}
const foo = ({
  a
}) => {}
const foo = ({
  a
}) => {}
",
      "warningCount": 0,
    },
  ],
  "warningCount": 0,
}
`;
